tools/ci/api-linter/linter/api-linter.go

```diff
@@ -0,0 +1,245 @@
+package linter
+
+import (
+	"flag"
+	"fmt"
+	"go/ast"
+	"path/filepath"
+	"reflect"
+	"strings"
+
+	"golang.org/x/tools/go/analysis"
+
+	"github.com/kumahq/kuma/pkg/util/pointer"
+)
+
+var eql = func(a, b string) bool { return a == b }
+
+var excludedPackages = map[string]func(string, string) bool{
+	"testdata": eql,
+	"_test":    strings.HasSuffix,
+}
+
+var excludedFiles = map[string]func(string, string) bool{
+	"zz_generated": strings.Contains,
+	"validator":    strings.Contains,
+	"compare":      strings.Contains,
+}
+
+var debugLog *bool
+
+var Analyzer = &analysis.Analyzer{
+	Name:  "apilinter",
+	Doc:   "checks that struct fields follow proper serialization rules",
+	Run:   run,
+	Flags: flags(),
+}
+
+func flags() flag.FlagSet {
+	set := flag.NewFlagSet("", flag.ExitOnError)
+	debugLog = set.Bool("debugLog", false, "disable nolint checks")
+	return *set
+}
+
+func run(pass *analysis.Pass) (interface{}, error) {
+	for _, file := range pass.Files {
+		fileName := pass.Fset.File(file.Pos()).Name()
+		fileNameWithoutExtension := stripExtension(fileName)
+		if shouldExcludeResource(pass.Pkg.Path(), excludedPackages) || shouldExcludeResource(fileName, excludedFiles) {
+			continue
+		}
+		ast.Inspect(file, func(n ast.Node) bool {
+			typeSpec, ok := n.(*ast.TypeSpec)
+			if !ok {
+				return true
+			}
+
+			if strings.ToLower(typeSpec.Name.String()) != fileNameWithoutExtension {
+				if *debugLog {
+					fmt.Println("DEBUG: Skipping type", typeSpec.Name.String(), "in file", fileNameWithoutExtension)
+				}
+				return true
+			}
+
+			structType, ok := typeSpec.Type.(*ast.StructType)
+			if !ok {
+				return true
+			}
+
+			analyzeStructFields(pass, structType, typeSpec.Name.Name, "", false)
+
+			return false
+		})
+	}
+	return nil, nil
+}
+
+func shouldExcludeResource(name string, rules map[string]func(string, string) bool) bool {
+	for pattern, matchFunc := range rules {
+		if matchFunc(name, pattern) {
+			return true
+		}
+	}
+	return false
+}
+
+func analyzeStructFields(pass *analysis.Pass, structType *ast.StructType, structName string, parentPath string, isMergeable bool) {
+	for _, field := range structType.Fields.List {
+		fieldName := ""
+		if len(field.Names) > 0 {
+			fieldName = field.Names[0].Name
+		}
+
+		// Correctly append field names to the path
+		var fieldPath string
+		if parentPath == "" {
+			fieldPath = structName + "." + fieldName
+		} else {
+			fieldPath = parentPath + "." + fieldName
+		}
+
+		if *debugLog {
+			fmt.Println("DEBUG: Analyzing field", fieldPath)
+		}
+
+		// Handle pointers to structs (*Struct)
+		baseType := field.Type
+		if ptrType, ok := field.Type.(*ast.StarExpr); ok {
+			baseType = ptrType.X // Unwrap pointer
+		}
+
+		// Recursively analyze named nested structs
+		if ident, ok := baseType.(*ast.Ident); ok {
+			namedStruct := findStructByName(pass, ident.Name)
+			if namedStruct != nil {
+				analyzeStructFields(pass, namedStruct, ident.Name, fieldPath, isMergeable)
```

Comment:
I'm probably missing something, but you're always passing `isMergeable` that came as an argument without modifying it. So it is probably always `false`.

Reply:
No, I was missing an if to check if it's under "Conf". Let me correct that.